cleaner.go

```diff
@@ -0,0 +1,362 @@
+	// correct for that type of id. This function can be nil, in which case no massaging will be done
+	massageId func(interface{}) interface{}
+}
+
+func (stats CollectionStats) Details() string {
+	return fmt.Sprintf("processed %d documents, removed %d, updated %d (%d tokens)\n"+
+		"checked %d tokens (%d completed unique) across %d completed transactions",
+		stats.DocCount, stats.RemovedCount, stats.UpdatedDocCount, stats.PulledTokenCount,
+		stats.TokenCount, stats.CompletedTokenCount, stats.CompletedTxnCount)
+}
+
+// stashId is special from other documents. When looking up a document in Mongo, the *order* of the attributes matter,
+// so we have to use a struct with defined ordering, instead of just holding it in a map with undefined ordering.
+type stashId struct {
+	C  string      `bson:"c"`
+	Id interface{} `bson:"id"`
+}
+
+// massageStashId turns a map[string]interface into a stashId if it matches.
+func massageStashId(id interface{}) interface{} {
+	m, ok := id.(bson.M)
+	if !ok {
+		logger.Warningf("massage failed, not a bson.M: %T %v", id, id)
+		return id
+	}
+	c, ok := m["c"]
+	if !ok {
+		logger.Warningf("massage failed, no 'c' field: %v", id)
+		return id
+	}
+	cStr, ok := c.(string)
+	if !ok {
+		logger.Warningf("massage failed, 'c' not a string: %v", id)
+		return id
+	}
+	collId, ok := m["id"]
+	if !ok {
+		logger.Warningf("massage failed, no 'id' field: %v", id)
+		return id
+	}
+	return stashId{
+		C:  cStr,
+		Id: collId,
+	}
+}
+
+// NewCollectionCleaner creates an object that can remove transaction tokens from document queues when the transactions
+// have been marked as completed.
+func NewCollectionCleaner(config CollectionConfig) *collectionCleaner {
+	if config.NumBatchTokens == 0 {
+		config.NumBatchTokens = queueBatchSize
+	}
+	if config.MaxRemoveQueue == 0 {
+		config.MaxRemoveQueue = maxMemoryTokens
+	}
+	return &collectionCleaner{
+		config:         config,
+		docIdsToRemove: make([]interface{}, 0),
+		docsToProcess:  make([]txnDocument, 0),
+		tokensToLookup: make([]string, 0),
+		removeIfEmpty:  false,
+		massageId:      nil,
+	}
+}
+
+// NewStashCleaner returns an object suitable for cleaning up all of the txns.stash collection.
+// It is unique in that the document Ids in this collection are special, and if we find all references from a document
+// have been removed, we no longer need the document to exist.
+func NewStashCleaner(config CollectionConfig) *collectionCleaner {
+	return &collectionCleaner{
+		config:         config,
+		docIdsToRemove: make([]interface{}, 0),
+		docsToProcess:  make([]txnDocument, 0),
+		tokensToLookup: make([]string, 0),
+		removeIfEmpty:  true,
+		massageId:      massageStashId,
+	}
+}
+
+// includeDoc queues this doc to be processed. It returns 'true' if the docs should be processed.
+func (cleaner *collectionCleaner) includeDoc(doc txnDocument) (bool, error) {
+	cleaner.stats.DocCount++
+	if cleaner.stats.DocCount%50000 == 0 {
```

Comment:
Consider using `simpleTimer` (in this package) to emit progress on a time basis instead of by document. Progress reporting isn't then tied to the performance of the machine. Switching to this for the other pruning code proved to work out really well.